File: cpp/test_baseline.cpp
#include <iostream>
#include <memory>
#include <fstream>
#include <inttypes.h>
#include <vector>
#include <iomanip>
#include "YKSConverter/YKSConverter.h"

void writeBufferToFile(std::shared_ptr<YKS::ByteBuffer> buffer, const std::string& filename) {
    std::ofstream out(filename, std::ios::binary);
    for (int i = 0; i < buffer->size(); i++) {
        out << buffer->get(i);
    }
    out.close();
    std::cout << "Generated: " << filename << " (size: " << buffer->size() << " bytes)" << std::endl;
}

void printBufferHex(std::shared_ptr<YKS::ByteBuffer> buffer, const std::string& name) {
    std::cout << name << " hex dump:" << std::endl;
    for (int i = 0; i < buffer->size(); i++) {
        std::cout << std::hex << std::setfill('0') << std::setw(2) << (int)buffer->get(i) << " ";
        if ((i + 1) % 16 == 0) std::cout << std::endl;
    }
    std::cout << std::dec << std::endl << std::endl;
}

int main() {
    std::cout << "=== YKSConverter C++ Baseline Tests ===" << std::endl;
    
    // Test Case 1: Simple single track
    std::cout << "Test 1: Simple single track" << std::endl;
    std::string mml1 = "MML@t120l4cdefgab>c4.,,,;";
    auto converter1 = std::make_shared<YKSConverter>(mml1, 1);
    auto buffer1 = converter1->toBuffer();
    if (buffer1) {
        writeBufferToFile(buffer1, "test1_simple_single.midi");
        printBufferHex(buffer1, "Test1");
    }
    
    // Test Case 2: Multi-track ensemble
    std::cout << "Test 2: Multi-track ensemble" << std::endl;
    std::vector<std::string> mml2 = {
        "MML@t180l8ccccccc4,l8eeeeeee4,l8ggggggg4;",
        "MML@t180l8>ccccccc4,l8>eeeeeee4,l8>ggggggg4;"
    };
    std::vector<uint8_t> inst2 = {26, 74};
    auto converter2 = std::make_shared<YKSConverter>(mml2, inst2);
    auto buffer2 = converter2->toBuffer();
    if (buffer2) {
        writeBufferToFile(buffer2, "test2_multi_track.midi");
        printBufferHex(buffer2, "Test2");
    }
    
    // Test Case 3: Complex MML from README
    std::cout << "Test 3: README example" << std::endl;
    std::string mml3 = "MML@t190l8cdefgab>c4.,l8<cdefgab>c4.,l8>cdefgab>c4.;";
    auto converter3 = std::make_shared<YKSConverter>(mml3, 1);
    auto buffer3 = converter3->toBuffer();
    if (buffer3) {
        writeBufferToFile(buffer3, "test3_readme_example.midi");
        printBufferHex(buffer3, "Test3");
    }
    
    // Test Case 4: Edge case - empty tracks
    std::cout << "Test 4: Edge case with empty tracks" << std::endl;
    std::string mml4 = "MML@t120l4cde,,;";
    auto converter4 = std::make_shared<YKSConverter>(mml4, 1);
    auto buffer4 = converter4->toBuffer();
    if (buffer4) {
        writeBufferToFile(buffer4, "test4_empty_tracks.midi");
        printBufferHex(buffer4, "Test4");
    }
    
    std::cout << "=== Baseline tests completed ===" << std::endl;
    return 0;
}

File: rust/src/bin/test_baseline.rs
use std::fs::File;
use std::io::Write;
use yks_converter::{YksConverter, ByteBuffer};

fn write_buffer_to_file(buffer: &ByteBuffer, filename: &str) -> Result<(), std::io::Error> {
    let mut file = File::create(filename)?;
    file.write_all(buffer.as_slice())?;
    println!("Generated: {} (size: {} bytes)", filename, buffer.size());
    Ok(())
}

fn print_buffer_hex(buffer: &ByteBuffer, name: &str) {
    println!("{} hex dump:", name);
    for (i, byte) in buffer.as_slice().iter().enumerate() {
        print!("{:02x} ", byte);
        if (i + 1) % 16 == 0 {
            println!();
        }
    }
    println!();
    println!();
}

fn main() {
    println!("=== YKSConverter Rust Baseline Tests ===");
    
    // Test Case 1: Simple single track
    println!("Test 1: Simple single track");
    let mml1 = "MML@t120l4cdefgab>c4.,,;".to_string();
    let converter1 = YksConverter::new(mml1, 1);
    if let Some(buffer1) = converter1.to_buffer() {
        write_buffer_to_file(&buffer1, "test1_simple_single_rust.midi").unwrap();
        print_buffer_hex(&buffer1, "Test1_Rust");
    }
    
    // Test Case 2: Multi-track ensemble  
    println!("Test 2: Multi-track ensemble");
    let mml2 = vec![
        "MML@t180l8ccccccc4,l8eeeeeee4,l8ggggggg4;".to_string(),
        "MML@t180l8>ccccccc4,l8>eeeeeee4,l8>ggggggg4;".to_string(),
    ];
    let inst2 = vec![26, 74];
    let converter2 = YksConverter::new_multi(mml2, inst2);
    if let Some(buffer2) = converter2.to_buffer() {
        write_buffer_to_file(&buffer2, "test2_multi_track_rust.midi").unwrap();
        print_buffer_hex(&buffer2, "Test2_Rust");
    }
    
    // Test Case 3: Complex MML from README
    println!("Test 3: README example");
    let mml3 = "MML@t190l8cdefgab>c4.,l8<cdefgab>c4.,l8>cdefgab>c4.;".to_string();
    let converter3 = YksConverter::new(mml3, 1);
    if let Some(buffer3) = converter3.to_buffer() {
        write_buffer_to_file(&buffer3, "test3_readme_example_rust.midi").unwrap();
        print_buffer_hex(&buffer3, "Test3_Rust");
    }
    
    // Test Case 4: Edge case - empty tracks
    println!("Test 4: Edge case with empty tracks");
    let mml4 = "MML@t120l4cde,,;".to_string();
    let converter4 = YksConverter::new(mml4, 1);
    if let Some(buffer4) = converter4.to_buffer() {
        write_buffer_to_file(&buffer4, "test4_empty_tracks_rust.midi").unwrap();
        print_buffer_hex(&buffer4, "Test4_Rust");
    }
    
    println!("=== Rust baseline tests completed ===");
}